inference-api/inference_api_python_client/client.py

```diff
@@ -0,0 +1,147 @@
+      client_private_key=client_private_key,
+      base_url=base_url,
+      auth_path=self.AUTH_PATH,
+    )
+
+  def _validate_args(
+    self,
+    private_key_filename: Optional[str] = None,
+    private_key_string: Optional[str] = None,
+  ) -> None:
+    if not private_key_filename and not private_key_string:
+      raise InvalidInferenceAPIClientArgsException(
+        reason='either private_key_filename or private_key_string must be provided')
+    if private_key_filename and private_key_string:
+      raise InvalidInferenceAPIClientArgsException(
+        reason='private_key_filename and private_key_string cannot both be provided')
+
+  async def get_inferences_async(
+    self,
+    model_id: str,
+    patient_mrn: str,
+    segmentation_destination_dirname: str,
+  ) -> List[Inference]:
+    """Asynchronous function to fetch inferences from the API matching the model_id and patient_mrn, and
+    download their segmentations locally.
+
+    Args:
+        model_id (str): Model ID for the inferences.
+        patient_mrn (str): Medical record number (MRN) of the patient.
+        segmentation_destination_dirname (str): Directory name where the segmentations will be downloaded.
+
+    Returns:
+        List[Inference]: List of Inference objects fetched from the API.
+    """
+    resource_path = self.GET_INFERENCE_PATH.format(
+      model_id=model_id,
+      patient_mrn=patient_mrn,
+    )
+    response_json = await self._django_jwt_client.get_json(resource_path)
+    inferences: List[Inference] = []
+    for inference in response_json:
+      await self._download_segmentations(
+        presigned_urls=inference['segmentation_presigned_urls'],
+        destination_dirname=segmentation_destination_dirname,
+      )
+      inferences.append(
+        Inference(
+          model_id=model_id,
+          patient_mrn=patient_mrn,
+          segmentation_presigned_urls=inference['segmentation_presigned_urls'],
+      ))
+    return inferences
+
+  def get_inferences(
+    self,
+    model_id: str,
+    patient_mrn: str,
+    segmentation_destination_dirname: str,
+  ) -> List[Inference]:
+    """Function to fetch inferences from the API matching the model_id and patient_mrn, and
+    download their segmentations locally.
+
+    Args:
+        model_id (str): Model ID for the inferences.
+        patient_mrn (str): Medical record number (MRN) of the patient.
+        segmentation_destination_dirname (str): Directory name where the segmentations will be downloaded.
+
+    Returns:
+        List[Inference]: List of Inference objects fetched from the API.
+    """
+    return async_to_sync(self.get_inferences_async)(
+      model_id=model_id,
+      patient_mrn=patient_mrn,
+      segmentation_destination_dirname=segmentation_destination_dirname,
+    )
+
+  async def _download_segmentations(
+    self,
+    presigned_urls: str,
+    destination_dirname: str,
+  ) -> None:
+    async with aiohttp.ClientSession() as session:
```

Comment:
The `aiohttp` guide advises against creating a new `ClientSession` for each request.

"Note: Don’t create a session per request. Most likely you need a session per application which performs all requests altogether." - https://docs.aiohttp.org/en/latest/client_quickstart.html

"Even though creating a session on demand seems like a tempting idea, we advise against it. [aiohttp.ClientSession](https://docs.aiohttp.org/en/latest/client_reference.html#aiohttp.ClientSession) maintains a connection pool. Contained connections can be reused if necessary to gain some performance improvements. If you plan on reusing the session, a.k.a. creating persistent session, you can use either [Signals](https://docs.aiohttp.org/en/latest/web_advanced.html#aiohttp-web-signals) or [Cleanup Context](https://docs.aiohttp.org/en/latest/web_advanced.html#aiohttp-web-cleanup-ctx). If possible we advise using [Cleanup Context](https://docs.aiohttp.org/en/latest/web_advanced.html#aiohttp-web-cleanup-ctx), as it results in more compact code:" - https://docs.aiohttp.org/en/latest/client_advanced.html#aiohttp-persistent-session

Reply:
Is it reasonable to assume or mandate that `InferenceAPIClient` will always be created in an asynchronous context? This could be as simple as requiring that the client application is started via `asyncio.run(main())`.

Reply:
I addressed this by making the `InferenceAPIClient` itself into a context manager, so that the session is now managed by the user of the client rather than the client itself, allowing for reused connections.

However, this is committing the client to an async direction - these changes make it so that the synchronous version no longer works. To make it work would require creating a parallel implementation of most of the functionality using a synchronous library like `requests`.

I'm hoping that maybe we can commit to either sync or async without having to maintain both.

@archerbh

Reply:
I think it's reasonable to proceed in the async direction. `asyncio` was introduced in Python 3.4, and asynchronous code should be standard practice for network calls.